firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java

```diff
@@ -50,15 +74,63 @@ static synchronized HeartBeatInfoStorage getInstance(Context applicationContext)
     return instance;
   }
 
+  synchronized void storeHeartBeatInformation(String heartBeatTag, long millis) {
+    this.heartBeatSharedPreferences.edit().putString(String.valueOf(millis), heartBeatTag).apply();
+    this.heartBeatCount += 1;
+    if (this.heartBeatCount > HEART_BEAT_COUNT_LIMIT) {
+      this.cleanUpStoredHeartBeats();
+    }
+  }
+
+  private synchronized void cleanUpStoredHeartBeats() {
+    ArrayList<Long> timestampList = new ArrayList<>();
+    for (Map.Entry<String, ?> entry : heartBeatSharedPreferences.getAll().entrySet()) {
+      timestampList.add(Long.parseLong(entry.getKey()));
+    }
+    Collections.sort(timestampList);
+    for (Long millis : timestampList) {
+      this.heartBeatSharedPreferences.edit().remove(String.valueOf(millis)).apply();
+      this.heartBeatCount -= 1;
+      if (this.heartBeatCount <= (HEART_BEAT_COUNT_LIMIT / 2)) return;
+    }
+  }
+
+  synchronized long getLastGlobalHeartBeat() {
+    return sharedPreferences.getLong(GLOBAL, -1);
+  }
+
+  synchronized void updateGlobalHeartBeat(long millis) {
+    sharedPreferences.edit().putLong(GLOBAL, millis).apply();
+  }
+
+  synchronized List<SdkHeartBeatResult> getStoredHeartBeats(boolean shouldClear) {
+    ArrayList<SdkHeartBeatResult> sdkHeartBeatResults = new ArrayList<>();
+    for (Map.Entry<String, ?> entry : heartBeatSharedPreferences.getAll().entrySet()) {
+      long millis = Long.parseLong(entry.getKey());
+      String sdkName = ((String) entry.getValue());
+      sdkHeartBeatResults.add(SdkHeartBeatResult.create(sdkName, millis));
+    }
+    Collections.sort(sdkHeartBeatResults);
+    if (shouldClear) clearStoredHeartBeats();
+    return sdkHeartBeatResults;
+  }
+
+  void clearStoredHeartBeats() {
+    heartBeatSharedPreferences.edit().clear().apply();
+  }
+
+  boolean isValidHeartBeat(long base, long target) {
+    return !((new Date(base).getDate()) == (new Date(target).getDate()));
```

Comment:
I wonder what will happen if the device changes timezone, can we use utc here? Additionally getDate() seems deprecated.

Also is it possible that the date is the same but the month is different?

Reply:
no the date would include a combination of month day and year. Yeah can use utc

Reply:
what I mean is that `date.getDate()` returns the day of the month, so there is no info about which month it is or which year

Reply:
Ah okay I thought it handled all the 3. I will change it.

Reply:
Regarding the first question, it shouldn't matter if the device changes timezone it should still be utc bcz we are getting the date based on the milliseconds which are the number of milliseconds since epoch (System.getMillis)